혼자 연습하기 로딩 및 연습 페이지 제작

voidmatcha opened the pull request:
> resolve #70

## 가져와야하는 상태
- 환결설정: 시간 초, 기업 이름, 직무 이름
- 선택한 질문 리스트 및 답변 - API를 통해 받아옴(Question, Answer, Order, Date)

## hooks
- [x] useWindowsSize

## 컴포넌트
- [x] 배경화면 추가
- [x] [TextBox](https://yapp-17th.github.io/Web_2_Client/feat/70/?path=/story/textbox--controls)
- [x] [AnswerBox](https://yapp-17th.github.io/Web_2_Client/feat/70/?path=/story/answerbox--answer-box-default)
- [x] [QuestionTextBox](https://yapp-17th.github.io/Web_2_Client/feat/70/?path=/story/questiontextbox--question-text-box-default)

## 구현해야하는 세부 페이지
- [x] 기본 질문을 시작 -> Trigger: 확인 버튼
- [x] 설명 1, 2 -> Trigger: setTimeout
- [x] 설명 3 -> Trigger: 바로 시작하기 버튼
- [x] 질문 및 답변 -> Trigger: 답변 제출 버튼 & setTimeout(상태 변경)
  - [x] 스크립트 toggle

## Logic
- [x] 질문 리스트 API 연동
- [ ] 앞페이지에서 시간, 이름, 직무 상태 들고오도록 수정 <- 추후 진행
- [ ] 다음페이지를 위해 녹화시 질문 끝나면 구간 저장 로직 추가 <- 추후 진행
- [x] 다음페이지로 넘어가는 로직 추가(history.push 사용) 
  - [x] 동영상 저장 blob url 전달 및 작동 여부 판별(정상작동함)

https://yapp-17th.github.io/Web_2_Client/feat/70/?path=/story/page-selftrainpage--self-train

Lavegaa committed:
feat: 질문 리스트 페이지 제작 (#66)

Lavegaa committed:
feat: 질문리스트 제작중 (#66)

Lavegaa committed:
feat: 리스트 api 연동 (#66)

Lavegaa committed:
feat: 리스트 컴포넌트 제작중 (#66)

Lavegaa committed:
feat: 질문 리스트 api 연동 완성 (#66)
TODO
상세 페이지로 넘어가는 router설정

Lavegaa committed:
feat: 질문 페이지 제작중 (#66)

Lavegaa committed:
feat: 질문 작성 페이지 제작 중 (#66)
상세 입력 페이지 질문 없는경우 제작

voidmatcha committed:
feat: 캠 상단 TextBox 컴포넌트 제작 완료 (#70)

- topText, bottomText Props를 넣어서 와이어프레임대로 표시가능

voidmatcha committed:
fix: Button 컴포넌트 호버시 opacity 효과를 주는 CSS 추가

- 버튼 hover할 때 피드백이 없어 추가

voidmatcha committed:
refactor: CamView 컴포넌트 video의 control 속성 제거

- 기능상 control attribute가 필요 없음

voidmatcha committed:
refactor: SideBar width 영역 축소 이슈 방지 css 추가

- min-width 속성을 줌

voidmatcha committed:
feat: window의 높낮이 useWindowSize Custom Hooks 추가

- window의 width와 height을 가져올 수 있는 기능 제공

voidmatcha committed:
feat: SelfTrainPage 질의응답 전까지 완성 (#70)

- step 상태를 나눠서 단계별로 페이지 이동 할  수 있도록 제작

voidmatcha commented on src/pages/SelfTrainPage/SelfTrainPage.js:
```diff
@@ -0,0 +1,140 @@
+`;
+
+const WrapContainer = styled.div`
+  height: ${({ height }) => height}px;
```
이부분 `height: 100%;`가 안먹어서 이렇게 처리했는데 혹시 다른 방법 아시나요..?

Lavegaa replied:
높이를 페이지의 전체를 주고싶으신건가요?
height: 100vh; 를 하면 페이지 전체 높이를 가져갑니다.
https://taimouse.tistory.com/8

voidmatcha commented on src/pages/SelfTrainPage/SelfTrainPage.js:
```diff
@@ -0,0 +1,140 @@
+    bottom: '3분 동안 예상 면접 질문에 대해 답변을 생각해 보세요.',
+    button: '바로 시작하기',
+  },
+];
```
이부분은 fixture로 별도 파일로 빼놓겠습니다.

voidmatcha commented on src/pages/SelfTrainPage/SelfTrainPage.js:
```diff
@@ -0,0 +1,140 @@
+  }, [step]);
+
+  return (
+    <Wrapper>
+      {step === STEP_FIRST ? <Sidebar /> : <></>}
+      <WrapContainer height={height}>
+        <WrapTop>
+          {step === STEP_FIRST ? (
+            <ProfileMenuContainer />
+          ) : (
+            <Icon type="cancel_circle" func={handleCancel} alt="cancel" />
+          )}
+        </WrapTop>
+        <WrapMiddle>
+          <TextBox
+            topText={
+              (step === STEP_LOADING_2 ? `${name}님은 ${company} ` : '')
+              + Fixture[step].top
+            }
+            bottomText={Fixture[step].bottom}
+          />
+          <WrapCamView>
+            <CamView height={width / 3} width={width / 1.5} />
+          </WrapCamView>
+          {Fixture[step].button ? (
+            <Button
+              theme="blue"
+              text={Fixture[step].button}
+              func={handleNext}
+            />
+          ) : (
+            <></>
+          )}
+        </WrapMiddle>
+      </WrapContainer>
+    </Wrapper>
```
회의때 언급했던 부분인데 이렇게 step을 나눠서 트랜지션을 만드는게 옳은 방법일까요?

Lavegaa replied:
음..좀 복잡해 보이긴 하네요 생각해보겠습니다 허허..

SkynI25 replied:
```jsx
import EmailPasswordForm from '@/pages/SignUp/EmailPasswordForm';
import ProfileForm from '@/pages/SignUp/ProfileForm';
import { STEPS } from '@/pages/SignUp/helpers';

const renderForm = (step, setStep) => {
  switch (step) {
    case STEPS.EMAIL_PASSWORD:
      return <EmailPasswordForm setStep={setStep} />;
    case STEPS.PROFILE:
      return <ProfileForm setStep={setStep} />;
  }
};

function SignUp() {
  const [step, setStep] = useState(STEPS.EMAIL_PASSWORD);

  return (
    <div className="signup container">
      <h1 className="text-center">계정 만들기</h1>
      {renderForm(step, setStep)}
      ...
   </div>
}
```
제가 지난번에 말씀드렸던 방식인데, 말씀하셨던 것처럼 컴포넌트 파일을 따로 선언해주고 가져와 사용하기 때문에 그에 따른 파일이 추가되긴 합니다.
지금 구현하는 방식에서 switch...case 문법으로 공통되는 부분을 모아서 렌더링 해주도록 하면 어떨까요?

dididy replied:
컴포넌트에 적용된 삼한연산자를 1개만 남기고 다 숏서킷 평가 하는 방식으로 쪼금 덜 복잡하게 수정했습니다.  

기태님이 제시해주신 방식을 하나남은 컴포넌트 삼항연산자부분에 적용해보겠습니다. 감사합니다!

voidmatcha replied:
트렌지션 자체가 나누기 애매한 부분도 있고 상함연산자 부분에 적용하려고 보니 전달해야 할 props가 생각보다 많아서.. 제 나름대로  가독성을 위한 정리를 해놨습니다.

https://github.com/Yapp-17th/Web_2_Client/pull/71/files#diff-9cbbb696c9202a8da5f3403929a8d457165353b4233c1b9de63cb439a10824e4R107-R187

voidmatcha commented on src/components/Button/Button.js:
```diff
@@ -11,6 +11,9 @@ const Box = styled.div`
   height: 60px;
   border-radius: 10px;
   user-select: none;
+  &: hover {
+    opacity: 70%;
+  }
```
button hover시 피드백이 없어서 hover할 때 opacity를 주는 css 코드를 추가했습니다.

voidmatcha committed:
hotfix: TextBox 컴포넌트 경로 문제 해결

- 수정된 부분 저장을 안하고 커밋함

voidmatcha committed:
feat: SelfTrainPage 로 history.push버튼, Route exact옵션 추가

- 추후에 지울 예정 배포시 제대로 동작하는지 확인하는 용도

voidmatcha committed:
feat: AnswerBox 컴포넌트 퍼블리싱 완성

- [ ] 추후 질문리스트 생성/수정 날짜 반영된 API 적용
- position absolute에서 관계를 만들려면 상위 태그에 relative를 해주어야 함
- literal template 줄바꿈 적용할려면 textarea를 사용해야 함
- ::-webkit-scrollbar 를 이용해 스크롤바 없앨 수 있음

voidmatcha committed:
refactor: RemainTime 컴포넌트 폴더명에 .js 제거

- 폴더명에 적절하지 않은 텍스트가 들어가있어서 이를 제거함

voidmatcha committed:
feat: QuestionTextBox 컴포넌트 제작 완성

- 혼자연습하기 시작 후 질문을 표시하는 컴포넌트 제작 완료

voidmatcha committed:
refactor: ToggleButton funcActive, funcDeactive props 추가

- 컴포넌트의 input이 checked일 때 실행할 함수, 아닐 때 실행할 함수에 대한 props 추가 및 실행 함수 추가

voidmatcha committed:
refactor: CamView 컴포넌트 status props isRequire 제거

- 녹화가 고려되지 않는 상황이 있을 수 있어서 기본값 설정

voidmatcha committed:
refactor: useMediaRecorder hooks intervalId state에서 useRef로

- interval id 저장하는데  상태 tracking할 필요가 없어서 useRef로 대체

voidmatcha committed:
fix: AnswerBox 컴포넌트 하단 blur 영역 css 수정

- blur영역이 고정이 안되는 문제가 있어 이를 수정함

voidmatcha committed:
refactor: QuestionTextBox 컴포넌트 PropTypes 타입 정상화

- 해당 컴포넌트의 width가 string인지를 판별하고 있는 PropTypes를 number로 수정

voidmatcha committed:
feat: 혼자 연습하기 로딩 및 연습 페이지 퍼블리싱 완성 (#70)

- [ ] 질문 리스트 API 연동
- [ ] 앞페이지에서 시간, 이름, 직무 상태 들고오도록 수정
- [ ] 다음페이지를 위해 구간 저장 로직 추가
- [ ] 다음 페이지로 넘어가는 로직 추가(history.push 쓸 듯)

- Fixture, Style 분리
- 세부 페이지 트렌지션 구현 완료

voidmatcha committed:
env: storybook panel 비활성화를 위해 manage.js에 옵션 추가

- addons를 @storybook/addons 모듈에서 가져와서 setConfig 메서드에 shotPanel 옵션을 지정하여 panel을 비활성화 함

voidmatcha committed:
env: storybook 전역 폰트 적용 및 CSS 리셋 추가

- 전역폰트 적용이 안되서 webpack-dev-server로 실행했을 때랑 다른 부분이 있어 적용
- CSS 리셋은 적용이 안되는 것 같음

voidmatcha committed:
refactor: Sidebar 컴포넌트 width 변동 문제 CSS 해결

- min-width, max-width를 지정해줘 width를 고정하도록 함

voidmatcha committed:
feat: SlefTranPage 스토리북 추가 및 Page 카테고리 지정 (#70)

- title에 /로 구분해서 카테고리화 가능 해당 스토리북은 Page 카테고리로 모아둠
- AlongQuestionCheckList 페이지도 /로 구분해서 Page 카테고리로 들어가도록 수정

voidmatcha committed:
Merge branch 'feat/#66' into feat/#70

voidmatcha committed:
feat: getQuestionItemAPI 연동 및 Fixture 파일 별도 생성 (#70)

- 에러처리를 하기 위해 try catch 사용
  - response가 제대로 오지 않는 경우 Fixture 사용하도록 구성
- getQuestionItemAPI를 사용해서 API를 가져오도록 함
- [ ] 질문 id를 가져와서 적용하는 부분 추가해야 함

voidmatcha committed:
refactor: <></>을 가지고 있는 삼항연산자 && 로 교체

- 조건 에따라 표현해야할 컴포넌트가 1개라면 삼항연사자를 &&으로 교체할 수 있음

voidmatcha committed:
refactor: AnswerBox height props 값 추가

- isRequire props이면서 값이 필요한 부분이라 추가함

Lavegaa commented on src/pages/SelfTrainPage/SelfTrainPage.style.js:
```diff
@@ -0,0 +1,74 @@
+import styled from 'styled-components';
```
style이 길어지는 컴포넌트는 다음과 같이 따로 빼서 가는것도 괜찮을 것 같네요.
참고해보겠습니다.

voidmatcha committed:
feat: 사이드바, 프로필컨테이너 페이지와 독립 및 timeReducer 상태 추가 (#70)

- 사이드바와 프로필 컨테이너는 랜딩, 로그인,  연습페이지 외에 모든 페이지에서  표시되어야 함
  - 이를 분리하여 App.js 에 추가 후 스타일링 구성
  - ProfileMenuContainer에 absolute css 추가
  - Sidebar와 SelfTrainPage가 flex 관계를 잘 맺을 수 있도록 하는 css 추가

- TextBox와 QuestionTextBox 높이가 달라서 padding css 추가
- timeReducer에 SelfTrainPage의 상태를 관리하도록함

voidmatcha committed:
refactor(selfTrain): 로딩, 연습, 평가 유기적으로연동 (#70)

- App.js에 self-checklist용 AuthRoute 추가
- AloneQuewstionCheckList에서 취소버튼 누를 때 원점으로가도록
- SelfTrainPage 401 떳을 때 setLogout 적용
- 필요없는 WrapContainer 에 전달하는 height props 제거

voidmatcha committed:
refactor: SelfTrainPage 컴포넌트 코드 가독성을 위한 리팩토링

- 필요없는 부분 삭제
- 비교연산 결과 상수화 시켜서 적용
- 가독성을 위해 상수와 비교연산자로 숏서킷 처리

voidmatcha committed:
feat: SelfTrainPage 컴포넌트 배경 이미지 추가 (#70)

- 원하는 step만 배경이 보이도록 구현

voidmatcha committed:
feat(record): 영상 Upload 로직 구현 - 400 에러뜸 WIP (#72)

- 재호님이 구현해주신 API에 blob 전송하도록 구현 완료
  - spostViedoRepository 추가
  - content-type 헤더 선택할 수 있도록 변경
- 테스트를 해보면서 CORS, 500 에러는 해결하였지만 400 에러가 발생..
- 재호님의 수월한 테스트를 위해 커밋 푸시
- checklist 에서 영상 확인할 수 있도록 수정한 부분 포함됨

voidmatcha committed:
refactor: video post 관련 repository 이름 바꿈 및 get 추가

- getVideoApi를 호출 시 해당 계정의 녹화된 비디오 객체들을 가져옴
- 기존 이름에 post가 들어가 있었는데 이제 get도 쓰므로 requestVideoRepository로 이름을 바꿈

voidmatcha committed:
feat(redux): 연습 진행중일 때 상태 - train reducer 생성

- uploadedLocation은 녹화 완료되고 서버에서 해당 동영상 URL을 저장
  - 추후 checklist의 video source에 적용해야 함
- localBlob은 전체 녹화 Blob을 저장함
  - 현재 checklist  video source에 지정
- 기존 time 리듀서에서 가지고 있던 mediaUrl을 train reducer에서 관리하도록 함

voidmatcha committed:
refactor: 동영상 POST로 서버에 업로드 하는 부분 완성 (#72)

- selectedQnaId의 필요성을 느껴 auth reducer에 추가
- 혼자 연습은 내 캠화면만 녹화하면 되므로 useReactMediaRecorder hooks 옵션에 screen 대신 video 넣음

- AloneQuestionCheckList에 비디오 렌더링 부분 녹화된 화면 재생하도록 수정
- useMediaRecorder에 필요한 로직 추가 및 setInterval 주석처리(이건 실시간 전송을 하지 않는이상 사용 안하게될듯)

- 쓸모없는 주석 제거

voidmatcha committed:
feat: 질문리스트 API 작성시간 추가된 부분 연동되도록 적용

- modifiedAt 으로 접근해서 가져옴

voidmatcha commented on package.json:
```diff
@@ -67,6 +67,7 @@
   },
   "dependencies": {
     "@reduxjs/toolkit": "^1.4.0",
+    "await-timeout": "^1.1.1",
```
UI 처리에 setTimeout을 사용할 경우 취소하는 부분이 더러워져서 간결하게 하기 위해 추가했습니다.

Lavegaa replied:
오 라이브러리가 있었군요 다행입니다!

voidmatcha replied:
찾아보니 hooks를 만들어 깔끔하게 사용하는 방법도 있더라구요 https://github.com/beautifulinteractions/beautiful-react-hooks/blob/master/docs/useTimeout.md

나중에 시간 되면 적용해보겠습니다.

Lavegaa replied:
별별 훅스가 다있네요 허허...네넵 일단 마무리하고 리팩토링 다같이 하시죠!

voidmatcha commented on src/App.js:
```diff
@@ -1,19 +1,45 @@
+        <AuthRoute path="/questionlist" component={QuestionListPage} />
+        <Route path="/question/:id" component={QuestionPage} />
         <Route path="/room/:roomID" component={ConferenceRoom} />
+        <Wrapper>
+          {!toggleTrain && <Sidebar />}
+          {!toggleTrain && <ProfileMenuContainer name={name} />}
+          <AuthRoute exact path="/self-training" component={SelfTrainPage} />
+          <AuthRoute exact path="/self-checklist" component={AloneQuestionCheckList} />
+        </Wrapper>
```
실제 연습중 & 체크리스트 화면에서 사이드바, 프로필컨테이너가 렌더링 되지 않아야 하는데 이를 판별하게 해줄 수 있도록 redux로 만들어둔 상태입니다. time 리듀서에 추가되어 있습니다.

Lavegaa replied:
페이지에 따로 sidebar를 넣지 않아도 될까요?

voidmatcha replied:
네 로그인 되어있다면 우측 상단에 떠있는 profile container와 함께 전역으로 관리됩니다!

Lavegaa replied:
오호.. sidebar 따로 넣어뒀던 부분은 빼야겠네요! 감사합니다.

voidmatcha commented on src/context/serverContext.js:
```diff
@@ -2,16 +2,19 @@ import Axios from 'axios';
 
-const api = ({ url, type = 'get', param }) => Axios({
+const api = ({
+  url, type = 'get', param, contentType = 'application/json',
+}) => Axios({
   method: type,
   url: `${SERVER_URL}${url}`,
   data: param,
   headers: {
-    'Content-Type': 'application/json',
+    'Content-Type': contentType,
     'Access-Control-Allow-Origin': '*',
     'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE',
     'Access-Control-Allow-Headers': 'Origin,Accept,X-Requested-With,Content-Type,Access-Control-Request-Method,Access-Control-Request-Headers,Authorization',
   },
+  withCredentials: true,
 });
```
녹화된 것을 POST로 서버에 보낼 경우 헤더의 Content-Type이 'multipart/form-data'로 변경되어야 해서 변수화 하였습니다.

Lavegaa replied:
오 감사합니다. 영상쪽에서 분명 필요 할 것 같긴 했는데 허허

voidmatcha commented on src/hooks/useMediaRecorder.js:
```diff
@@ -127,21 +119,46 @@ export default function useReactMediaRecorder({
+
     onStop(url, blob);
 
+    // TODO: 혼자연습하기 체크리스트에서 재생할 수 있도록 해야 함
+    const blobData = new Blob(mediaChunks.current, {
+      type: 'video/webm;codecs=vp8,opus',
+    });
+    const formData = new FormData();
+    formData.append('videoFile', blobData);
+    formData.append('questionListId', selectedQnaId);
+
+    // TODO: 서버의 Response를 기다리는 로딩 상황임을 알려주는 Progress Bar 혹은 Spinner가 렌더링 되게끔 하는 상태를 추가해야 함
+    await postVideoApi(formData)
+      .then((response) => {
+        const { id } = response.data;
+        getVideoApi().then((res) => {
+          const { savedLocation } = res.data.find((elem) => elem.id === id);
+          const uploadedLocation = `${savedLocation.substring(0, 27)}videos/${savedLocation.substring(27)}`;
+          dispatch(setUploadedLocation({ uploadedLocation }));
+        });
+      })
+      .catch((err) => {
+        alert(err);
+      });
+
```
이부분이 서버로 녹화영상 blob을 업로드하는 부분입니다.

Lavegaa replied:
저번에 잠깐 이야기 들었을때는 blob으로 보낸 영상파일을 우리가 원하는 확장자로 저장이 가능한건가요? 재호님예은님께 여쭤봐야하는건가..ㅋㅋㅋ

voidmatcha replied:
그부분은 전부 해결되었고 local blob 재생도 잘 됩니다. 다만.. 서버에 저장된 hls url을 불러와서 재생하는것에는 실패했습니다. 일반적인 hls 재생 모듈에 url을 넣는 방식으로는 안되더라구요.. 재호님과 예은님이 샘플코드 링크를 주셨는데 이거대로 하면 가능할 것 같습니다.(아직 테스트는 안해봤어요)

https://github.com/Yapp-17th/Web_2_Client/issues/78 <- 여기 설명되어 있습니다.

Lavegaa replied:
참고해보겠습니다!

SkynI25 replied:
와 ..감사합니다.

voidmatcha commented on src/hooks/useWindowSize.js:
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+
+export default function useWindowSize() {
+  const isBrowser = typeof window !== 'undefined';
+  const [windowSize, setWindowSize] = React.useState({
+    width: isBrowser ? window.innerWidth : 1200,
+    height: isBrowser ? window.innerHeight : 590,
+  });
+
+  function changeWindowSize() {
+    setWindowSize({ width: window.innerWidth, height: window.innerHeight });
+  }
+
+  React.useEffect(() => {
+    window.addEventListener('resize', changeWindowSize);
+
+    return () => {
+      window.removeEventListener('resize', changeWindowSize);
+    };
+  }, []);
+
+  return windowSize;
+}
```
컴포넌트 구현시 현재 윈도우 높이와 넓이를 알아야 할 필요성이 있고 이를 간결하게 사용할 수 있도록 custom hooks로 만들었습니다.

voidmatcha commented on src/pages/AloneQuestionCheckList/AloneQuestionCheckList.js:
```diff
@@ -276,15 +281,18 @@ export default function AloneQuestionCheckList({ src }) {
     <Background>
       <CloseButton type="button">
-        <Icon type="cancel_circle" isCircle alt="close" />
+        <Icon type="cancel_circle" isCircle alt="close" func={() => history.push('/self-training')} />
```
기태님이 작성하셨던 부분인데 취소버튼을 누르면 앞 페이지로 이동하도록 처리하였습니다.

SkynI25 replied:
미처 처리못한 부분이었는데 감사합니다! 👍

voidmatcha commented on src/pages/AloneQuestionCheckList/AloneQuestionCheckList.js:
```diff
@@ -276,15 +281,18 @@ export default function AloneQuestionCheckList({ src }) {
-      <EndingTitle>면접이 종료 되었습니다. 체크리스트를 통해 스스로 평가를 해보세요.</EndingTitle>
+      <EndingTitle>
+        면접이 종료 되었습니다. 체크리스트를 통해 스스로 평가를 해보세요.
+      </EndingTitle>
       <Content>
         <LeftContent>
           <VideoContainer>
             <video ref={video} controls preload="metadata" onLoadedMetadata={loadVideoMetaData} onPlay={onPlay} onPause={onPause} onTimeUpdate={onTimeUpdate}>
               <track src="" kind="captions" srcLang="ko" label="korean_captions" />
-              <source src={src} type="video/webm" />
+              {/* TODO: 지금은 로컬 blob을 보여주고 있으나 이부분 서버 HLS url을 적용하는 방식으로 처리해야 함 */}
+              <source src={localBlob} type="video/mp4" />
```
이부분도 기태님이 작성하신 부분인데 앞서 useMediaRecorder에서 영상이 녹화영상 blob을 가지고 있는 상태 localBlob을 가져와 재생할 수 있도록 처리하였습니다. 
이부분은 HLS 플레이어로 교체하고 서버에서 영상 재생을 위해 제공하는 url을 가지고 있는 상태인 uploadedLocation변경해야 합니다.

SkynI25 replied:
그렇군요, 감사합니다!

voidmatcha commented on src/pages/SelfTrainPage/SelfTrainPage.js:
```diff
@@ -0,0 +1,212 @@
+    startRecording,
+    stopRecording,
+  } = useReactMediaRecorder({
+    video: true,
```
이부분은 1:1이 아닌경우 내 웹캠 영상만 저장하면 되기 때문에 screen에서 video로 변경하였습니다.

voidmatcha commented on src/pages/SelfTrainPage/SelfTrainPage.js:
```diff
@@ -0,0 +1,212 @@
+
+  const [questionList, setQuestionList] = useState(QNA_LIST);
+
+  const fetch = async (id) => {
+    try {
+      await getQuestionItemAPI(id).then((response) => {
+        setQuestionList(response.data);
+      });
+    } catch (err) {
+      if (err.response.status === 401) {
+        dispatch(setLogout());
+      }
+      setQuestionList(QNA_LIST);
+    }
+  };
```
질문 리스트를 가져오는 로직입니다. 401 status인 경우 로그아웃 되도록 하였습니다.

Lavegaa replied:
저도 401일때 로그아웃 되는 로직을 전체적으로 추가해야겠어요... auth router에서 한번에 처리하면 이상할까요?
페이지 이동시마다 로그인체크... 각각 api를 부를 때 하는게 더 정확하고 안전할까요

voidmatcha replied:
Auth router에서 처리한다는게 저는 감이 잘 안잡히는데 그부분 추가해주시면 비교해서 더 좋은 방식을 쓰면 될 것 같아요! 일일이 적용하는 방식은 아무래도 번거롭긴 해서..

SkynI25 replied:
저도 로그아웃 하는 부분을 질문체크리스트에 추가하겠습니다!

voidmatcha commented on src/pages/SelfTrainPage/SelfTrainPage.js:
```diff
@@ -0,0 +1,212 @@
+    }
+  };
+
+  useEffect(() => {
+    dispatch(handleReset({ keepTrain: false }));
+    fetch(selectedQnaId);
+  }, []);
```
렌더링 시점에 상태를 리셋하고 API 요청을 하도록 하는 부분 입니다.

voidmatcha commented on src/pages/SelfTrainPage/SelfTrainPage.js:
```diff
@@ -0,0 +1,212 @@
+    fetch(selectedQnaId);
+  }, []);
+
+  const handleCancel = async () => {
+    await stopRecording();
+    transition.clear();
+    dispatch(handleReset({ keepTrain: false }));
+  };
```
우측 상단 취소 버튼을 누르면 녹화중지, UI 트랜지션용 delay를 취소, 상태를 리셋하도록 하는 부분입니다.

voidmatcha commented on src/pages/SelfTrainPage/SelfTrainPage.js:
```diff
@@ -0,0 +1,212 @@
+    dispatch(handleReset({ keepTrain: false }));
+  };
+
+  const handleChecklistPage = async () => {
+    await stopRecording();
+    history.push('/self-checklist');
+    dispatch(handleReset({ keepTrain: true }));
+  };
```
기태님이 작성하신 페이지로 넘어가는 부분입니다. 녹화를 멈추고 상태를 리셋하는 과정이 있습니다.

voidmatcha commented on src/pages/SelfTrainPage/SelfTrainPage.js:
```diff
@@ -0,0 +1,212 @@
+    dispatch(handleReset({ keepTrain: true }));
+  };
+
+  useEffect(() => {
+    if (!time && step === STEP_START) {
+      dispatch(setStep({ step: STEP_ING }));
+      startRecording();
+    }
+
+    if (!time && (step === STEP_ING || step === TOGGLE_SCRIPT)) {
+      if (qnaStep === questionList.length - 1) handleChecklistPage();
+      else dispatch(handleStepQuestion());
+    }
+  }, [time]);
```
시간이 0초일 때 step에 따라 특정 함수를 실행하도록 하는 부분입니다.

voidmatcha commented on src/pages/SelfTrainPage/SelfTrainPage.js:
```diff
@@ -0,0 +1,212 @@
+    }
+  }, [time]);
+
+  useEffect(() => {
+    if (step === STEP_LOADING_1 || step === STEP_LOADING_2) {
+      transition.set(3000).then(() => dispatch(setStep({ step: step + 1 })));
+    }
+
+    if (step === STEP_START) {
+      dispatch(startTime({ count: 180 }));
+    }
+  }, [step]);
```
step에 따라 UI 트랜지션 부분이랑 준비시간을 설정하는 것을 실행하도록 하는 부분입니다.

voidmatcha commented on src/pages/SelfTrainPage/SelfTrainPage.js:
```diff
@@ -0,0 +1,212 @@
+    }
+  }, [step]);
+
+  const isStepFirst = step === STEP_FIRST;
+  const isLoading = step <= STEP_START;
+  const isBackground = step > STEP_FIRST && step < STEP_ING;
+  const isShowTimer = step >= STEP_START;
+  const isShowToggle = step > STEP_START;
+  const isShowAnswer = step === TOGGLE_SCRIPT;
+
+  const textBox = (
+    <TextBox
+      topText={
+        (step === STEP_LOADING_2 ? `${name}님은 ${COMPANY} ${JOB}` : '')
+        + Fixture[step]?.top
+      }
+      bottomText={Fixture[step]?.bottom || ''}
+    />
+  );
+
+  const questionTextBox = (
+    <QuestionTextBox
+      question={questionList[qnaStep]?.question || ''}
+      order={questionList[qnaStep]?.order + 1 || 0}
+      width={width / 1.5}
+    />
+  );
```
비교연산자의 결과를 변수로 빼고 삼항연산자로 처리해야 하는 jsx도 변수로 빼서 가독성을 향상시킬 수 있도록 하였습니다.

voidmatcha commented on src/pages/SelfTrainPage/SelfTrainPage.js:
```diff
@@ -0,0 +1,211 @@
+                answer={questionList[qnaStep].answer}
+                height={width / 3}
+                date={questionList[qnaStep].modifiedAt}
+              />
+            )}
+          </S.WrapCamView>
+          <S.WrapBottom width={width / 1.5}>
+            <S.WrapBottomSide>
+              {isShowTimer && <RemainTime time={time} />}
+            </S.WrapBottomSide>
+            {/* STEP_ING의 경우 버튼 누를 때 구간 저장하는 로직 추가 필요 */}
+            {Fixture[step]?.button && (
+              <Button
+                theme="blue"
+                text={Fixture[step].button}
+                func={
+                  qnaStep === questionList.length - 1
+                    ? () => handleChecklistPage()
+                    : () => {
+                      if (step === STEP_START) startRecording();
+                      dispatch(handleNextButton());
+                    }
+                }
+              />
+            )}
+            <S.WrapBottomSide right>
+              {isShowToggle && (
+                <>
+                  <S.WrapText>답변 보기 허용</S.WrapText>
+                  <ToggleButton
+                    funcActive={() => dispatch(setStep({ step: TOGGLE_SCRIPT }))}
+                    funcDeactive={() => dispatch(setStep({ step: STEP_ING }))}
+                  />
+                </>
+              )}
+            </S.WrapBottomSide>
+          </S.WrapBottom>
+        </S.WrapContent>
+      </S.WrapContainer>
+    </S.Wrapper>
```
이부분은 처음에 작성했을 때 보다 가독성 측면에서 잘 정리되었다고 생각합니다.

SkynI25 replied:
확실히 깔끔해졌네요. 외부 wrapping 없이 깔끔하게 작성할 수 있는 방법을 배웁니다. 정말 고생많으셨습니다.

voidmatcha commented on src/pages/SelfTrainPage/SelfTrainPage.js:
```diff
@@ -0,0 +1,211 @@
+            <S.WrapBottomSide>
+              {isShowTimer && <RemainTime time={time} />}
+            </S.WrapBottomSide>
+            {/* STEP_ING의 경우 버튼 누를 때 구간 저장하는 로직 추가 필요 */}
+            {Fixture[step]?.button && (
+              <Button
+                theme="blue"
+                text={Fixture[step].button}
+                func={
+                  qnaStep === questionList.length - 1
+                    ? () => handleChecklistPage()
+                    : () => {
+                      if (step === STEP_START) startRecording();
+                      dispatch(handleNextButton());
+                    }
+                }
+              />
+            )}
```
질문 시점 구간을 기록하는 부분을 추가해야 하는데 후순위로 두고 추후 작업하겠습니다.

voidmatcha commented on src/pages/SelfTrainPage/SelfTrainPage.js:
```diff
@@ -0,0 +1,211 @@
+              {isShowToggle && (
+                <>
+                  <S.WrapText>답변 보기 허용</S.WrapText>
+                  <ToggleButton
+                    funcActive={() => dispatch(setStep({ step: TOGGLE_SCRIPT }))}
+                    funcDeactive={() => dispatch(setStep({ step: STEP_ING }))}
+                  />
```
기태님이 작업해두신 토글 버튼에 on, off 상태에 따라 함수를 실행할 수 있는 props를 만들어서 적용하였습니다.

voidmatcha commented on src/components/ToggleButton/ToggleButton.stories.js:
```diff
@@ -7,3 +7,24 @@ export default {
 };
 
 export const Toggle = (args) => <ToggleButton {...args} />;
+export const ToggleWithFunction = (args) => {
+  const [state, setState] = useState('Deactive');
+
+  const handleClickActive = () => {
+    setState('Active');
+  };
+
+  const handleClickDeactive = () => {
+    setState('Decative');
+  };
+  return (
+    <>
+      <ToggleButton
+        {...args}
+        funcActive={handleClickActive}
+        funcDeactive={handleClickDeactive}
+      />
+      <div>{state}</div>
+    </>
+  );
+};
```
이부분이 토글 버튼 두가지 상태 각각 props로 들어오는 함수를 실행할 수 있도록 수정한 부분입니다.

voidmatcha commented on src/repository/requestVideoRepository.js:
```diff
@@ -0,0 +1,16 @@
+/* eslint-disable import/prefer-default-export */
+/* eslint-disable no-return-await */
+import api from '../context/serverContext';
+
+export const postVideoApi = async (param) => await api({
+  url: '/api/self/history',
+  type: 'post',
+  param,
+  contentType: 'multipart/form-data',
+});
+
+export const getVideoApi = async (param) => await api({
+  url: '/api/self/history',
+  type: 'get',
+  param,
+});
```
이부분은 동영상을 저장하고 현재 계정에 대해 저장된 영상 정보를 가져오는 것을 처리할 수 있도록 해 줍니다.

voidmatcha commented on src/store/Auth/auth.js:
```diff
@@ -4,26 +4,36 @@ const authReducer = createSlice({
-    user: '',
+    email: '',
+    name: '',
+    selectedQnaId: 3,
```
selectedQnaId는 다른 페이지 작업 하면서 train reducer로 관리하도록 할 계획입니다.

Lavegaa replied:
selectedQnaId가 나중에 train reducer로 옮긴다는 건가요?
어떤역할을 하는 state인가요??

voidmatcha replied:
네네 이거는 현재 선택된 질의응답묶음의 id를 저장하는 역할을 합니다. 중간에 API 요청할 때 이게 필요하라구요.. 서버로 영상 업로드 하는데에도 이것을 사용합니다. 
https://github.com/Yapp-17th/Web_2_Client/pull/71/files#diff-b4732697c380d1da9ca71e0cfed384f263af08c9e093433cb5094aec8aba4152R146

SkynI25 replied:
@dididy 혹시 default 값으로 3을 지정하신 이유가 있을까요?

voidmatcha commented on src/store/Time/time.js:
```diff
@@ -12,13 +16,109 @@ const timeReducer = createSlice({
+  const {
+    time: { qnaStep },
+  } = getState();
+  // TODO: 앞에서 설정한 TIME으로 변경해야 함
```
다른 페이지 작업 하면서 그 페이지의 상태를 설정했을 때 상태를 들고올 수 있도록 수정할 계획입니다.

voidmatcha commented on src/store/Train/train.js:
```diff
@@ -0,0 +1,28 @@
+const { actions, reducer } = createSlice({
+  name: 'application',
+  initialState: {
+    uploadedLocation: '',
+    localBlob: '',
```
uploadedLocation은 현재 업로드한 영상에 대해 서버에서 제공하는 url이고 localBlob은 로컬에 저장되어있는 녹화된 영상의 blob을의도하고 상태를 만들었습니다.

voidmatcha reviewed (commented):
변경점이 많고 찬호님 커밋도 일부 들어가 있어서 공유되어야 하는 부분에 대해서 코멘트를 해보았습니다!

Lavegaa commented on src/components/ConferenceButton/ConferenceButton.js:
```diff
@@ -20,6 +20,7 @@ export default function ConferenceButton({ history }) {
       <button type="button" onClick={() => handleCreate()}>컨퍼런스 생성</button>
       <button onClick={() => dispatch(showModal(MODALS.QUESTIONLIST_SAVE_MODAL))} type="button">Modal open</button>
       <button onClick={() => dispatch(hideModal(MODALS.QUESTIONLIST_SAVE_MODAL))} type="button">Modal close</button>
+      <button type="button" onClick={() => history.push('/self-training')}>혼자연습하기 로딩 및 질의응답</button>
```
react router의 Link태그와 history push중 어느걸 사용하는게 좋을까요?

https://medium.com/@han7096/react-router-v4-%EC%A0%95%EB%A6%AC-e9931b63dcae

위의 글을 보면 history.push는 match, history, location값들을 넘겨준다고 나오긴 하는데 Link는 안넘겨주려나... 잘 모르겠네요 이건

voidmatcha replied:
아마 js로 제어하기 위해서는 history.push를 사용하는게 맞는 것 같아요..!

Link는 jsx상에서 분기를 나누는 느낌인 것 같고.. 정확하진 않습니당

Lavegaa reviewed (approved):
헉 정말 고생 많으셨습니다!
영상관련 전체적으로 스터디를 한번 더 해봐야겠네요..대단하십니다👍

Commit:
Merge branch 'master' into feat/#70

SkynI25 reviewed (commented):
정말 많은 부분을 구현하셨네요..고생많으셨습니다!
